hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -79,6 +111,355 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+     * transfer the ownership of the NFTs from old treasury to new treasury
+     * @param oldTreasury old treasury account
+     * @param newTreasury new treasury account
+     * @param token token
+     * @param tokenRelStore token relationship store
+     * @param accountStore account store
+     */
+    private void transferTokensToNewTreasury(
+            final AccountID oldTreasury,
+            final AccountID newTreasury,
+            final Token token,
+            final WritableTokenRelationStore tokenRelStore,
+            final WritableAccountStore accountStore) {
+        final var tokenId = asToken(token.tokenNumber());
+        // Validate both accounts are not frozen and have the right keys
+        final var oldTreasuryRel = getIfUsable(oldTreasury, tokenId, tokenRelStore);
+        final var newTreasuryRel = getIfUsable(newTreasury, tokenId, tokenRelStore);
+        if (oldTreasuryRel.balance() > 0) {
+            validateFrozenAndKey(oldTreasuryRel);
+            validateFrozenAndKey(newTreasuryRel);
+
+            if (token.tokenType().equals(FUNGIBLE_COMMON)) {
+                // Transfers fungible balances and updates account's numOfPositiveBalances
+                // and puts to modifications on state.
+                transferFungibleTokens(oldTreasuryRel, newTreasuryRel, tokenRelStore, accountStore);
+            } else {
+                // Transfers NFT ownerships and updates account's numOwnedNfts and
+                // tokenRelation's balance and puts to modifications on state.
+                changeOwner(oldTreasuryRel, newTreasuryRel, tokenRelStore, accountStore);
+            }
+        }
+    }
+    /**
+     * Transfer fungible tokens from old treasury to new treasury.
+     * NOTE: This updates account's numOfPositiveBalances and puts to modifications on state.
+     * @param fromTreasuryRel old treasury relationship
+     * @param toTreasuryRel new treasury relationship
+     * @param tokenRelStore token relationship store
+     * @param accountStore account store
+     */
+    private void transferFungibleTokens(
+            final TokenRelation fromTreasuryRel,
+            final TokenRelation toTreasuryRel,
+            final WritableTokenRelationStore tokenRelStore,
+            final WritableAccountStore accountStore) {
+        final var adjustment = fromTreasuryRel.balance();
+
+        final var fromTreasury = accountStore.getAccountById(asAccount(fromTreasuryRel.accountNumber()));
+        final var toTreasury = accountStore.getAccountById(asAccount(toTreasuryRel.accountNumber()));
+
+        adjustBalance(fromTreasuryRel, fromTreasury, -adjustment, tokenRelStore, accountStore);
+        adjustBalance(toTreasuryRel, toTreasury, adjustment, tokenRelStore, accountStore);
+        // TODO: If any of the above fail, need to rollback only token transfer balances for record.
+        // Not sure how it will be done yet
```

Comment:
Could this be handled by the stackable states API?

Reply:
I am not sure yet.

Reply:
Me either. It sounds like it could, but I guess we'll see.